File: 卡码/代码随想录/1/06-二叉树/Go/03-二叉树修改与构造/LeetCode_0106.go
//106. 从中序与后序遍历序列构造二叉树
package main

import (
	. "binaryTree/types"
)

func traversal(inOrder []int,inLeft,inRight int,postOrder []int,postLeft,postRight int) *TreeNode{
	if inLeft>inRight {
		return nil
	}
	root:=&TreeNode{postOrder[postRight],nil,nil}
	if inLeft==inRight{
		return root
	}
	rootIndex:=0
	for i:=inLeft;i<=inRight;i++ {
		if inOrder[i]==root.Val {
			rootIndex=i
			break
		}
	}
	leftLength:=rootIndex-inLeft
	root.Left=traversal(inOrder,inLeft,rootIndex-1,postOrder,postLeft,postLeft+leftLength-1)
	root.Right=traversal(inOrder,rootIndex+1,inRight,postOrder,postLeft+leftLength,postRight-1)
	return root
}
func buildTree(inorder []int, postorder []int) *TreeNode {
	if len(inorder)==0 || len(postorder)==0 {
		return nil
	}
	return traversal(inorder,0,len(inorder)-1,postorder,0,len(postorder)-1)
}

File: 卡码/代码随想录/1/06-二叉树/Js/03-二叉树修改与构造/LeetCode_0106.js
//106. 从中序与后序遍历序列构造二叉树
import TreeNode from "../TreeNode.js";

const traversal=function(inOrder,inLeft,inRight,postOrder,postLeft,postRight){
  if(inLeft>inRight) return null;
  const root=new TreeNode(postOrder[postRight],null,null);
  if(inLeft==inRight) return root;
  let rootIndex=0;
  for(let i=inLeft;i<=inRight;i++){
    if(inOrder[i]==root.val){
      rootIndex=i;
      break;
    }
  }
  const leftLength=rootIndex-inLeft;
  root.left=traversal(inOrder,inLeft,rootIndex-1,postOrder,postLeft,postLeft+leftLength-1);
  root.right=traversal(inOrder,rootIndex+1,inRight,postOrder,postLeft+leftLength,postRight-1);
  return root;
}

const buildTree = function(inorder, postorder) {
  if(inorder.length==0||postorder.length==0) return null;
  return traversal(inorder,0,inorder.length-1,postorder,0,postorder.length-1);
};